File: src/etch/bindings/java/support/test/TestStubBase.java
package etch.bindings.java.support.test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.fail;

import org.junit.Test;

import metreos.core.Who;
import etch.bindings.java.msg.Message;
import etch.bindings.java.support.MessageSource;
import etch.bindings.java.support.StubBase;
import etch.bindings.java.support.Unwanted;

/** Test StubBase */
public class TestStubBase
{
	/** */
	@Test
	public void nothing()
	{
		// nothing to do.
	}
	
    /** test events */
	enum What
	{ 
	    /** wanted message delivered */ UNWANTED
	};
	
    /** @throws Exception */
	// TODO @Test
	public void unwanted2() throws Exception
	{
		MyUnwanted obj = new MyUnwanted();
		MessageSource src = new MyMessageSource();
		StubBase sb = new StubBase( src, obj, null, null );
		
		Who sender = new Who() { /* nothing */ };
		Message msg = new Message( null, null );
		
		sb.message( sender, msg );
		
		assertEquals( obj.what, What.UNWANTED );
		assertSame( sender, obj.xsender );
		assertSame( msg, obj.xmsg );
	}

    /** @throws Exception */
	// TODO @Test( expected=UnsupportedOperationException.class )
	public void unwanted4() throws Exception
	{
		Object obj = new Object();
		MessageSource src = new MyMessageSource();
		StubBase sb = new StubBase( src, obj, null, null );
		
		Who sender = new Who() { /* nothing */ };
		Message msg = new Message( null, null );
		
		sb.message( sender, msg );
		fail( "delivery of unwanted message should fail" );
	}
	
	/**
	 * MessageSource for testing.
	 */
	public static class MyMessageSource implements MessageSource
	{
		public void messagex( Who recipient, Message msg ) throws Exception
		{
			fail( "not needed" );
		}
	}
	
	/**
	 * Unwanted handler for testing.
	 */
	public static class MyUnwanted implements Unwanted
	{
		public void _unwanted( Who sender, Message msg )
		{
			assertNull( what );
			what = What.UNWANTED;
			this.xsender = sender;
			this.xmsg = msg;
		}
		
		/** what event */
		public What what;
		
		/** sender of the unwanted message */
		public Who xsender;
		
		/** unwanted message */
		public Message xmsg;
	}
}

File: src/etch/bindings/csharp/support/test/TestStubBase.cs
///$Id$
///
///Created by Champakesan, Badri Narayanan on Jul 02, 2007.
///
///Copyright (c) 2007 Cisco Systems, Inc. All rights reserved.

using System;
using NUnit.Framework;
using Etch.Msg;
using Etch.Support;
using Etch.Transport;

namespace Etch.Support.Test
{
    [TestFixture]
    public class TestStubBase
    {
        public enum What 
        { 
            /** wanted message delivered */ UNWANTED 
        };

        MyUnwanted obj = new MyUnwanted();

        [Test]
        public void Nothing()
        { }
        
        //[Test]
        public void Unwanted2()
        {
            MyUnwanted obj = new MyUnwanted();

            MessageSource src = new MyMessageSource();

            StubBase sb = new StubBase( src, obj, null, null );

            Who sender = new MyWho(); 
            Message msg = new Message(null, null);

            sb.Message( sender, msg);


            Assert.AreEqual( obj.what, What.UNWANTED );
            Assert.AreEqual( sender, obj.xsender );
            Assert.AreEqual( msg, obj.xmsg );
        }

        //[Test] 
        [ExpectedException("System.InvalidOperationException")]
        public void Unwanted3()
        {
            Object obj = new Object();

            MessageSource src = new MyMessageSource();
            StubBase sb = new StubBase( src, obj, null, null );

            Who sender = new MyWho(); 
            Message msg = new Message(null, null);

            sb.Message(sender, msg);
           Assert.Fail("delivery of unwanted message should fail");
        }

        //[Test]
        [ExpectedException("System.InvalidOperationException")]
        public void Unwanted4()
        {
            Object obj = new Object();

            MessageSource src = new MyMessageSource();
            StubBase sb = new StubBase( src, obj, null, null );

            Who sender = new MyWho(); 
            Message msg = new Message(null, null);

            sb.Message(sender, msg);
            Assert.Fail("delivery of unwanted message should fail");
        }

        public class MyMessageSource : MessageSource
        {

            #region MessageSource Members

            public void Messagex( Who recipient, Message msg )
            {
                Assert.Fail( "not needed" );
            }

            #endregion
        }

        public class MyUnwanted : Unwanted
        {
            public void Init()
            {
                what = null;
            }

            public void _Unwanted(MessageSource src, Who sender, Message msg)
            {
                Init();
                Assert.IsNull(what);
                what = What.UNWANTED;
                this.xsender = sender;
                this.xmsg = msg;
            }

            public Enum what;

            public Who xsender;

            public Message xmsg;
        }
    }

    public class MyWho : Who
    {
    }
}